Add `Pull request workflow` to contribution.md (#10672)
DataDog/datadog-agent: docs/dev/contributing.md

complete no-op for the test suite, please add **[skip ci]** in the commit message
body to skip the build and give that slot to someone else who does need it.

### Pull request workflow

The goals ordered by priority are:
- Make PR reviews (both initial and follow-up reviews) easy for reviewers using GitHub

> - On the `main` branch, have a meaningful commit history that allows understanding (even years later) what each commit does, and why. 

Reviewer: The lines in the patch should probably wrap at the same char limit as the rest of the file

Reviewer 2: Not sure to understand how we have a meaningful commit history as we use squash merged (which destroys the history and therefore merges a big PR into one single commit)

> #### Before the first in-depth PR review

Reviewer 3: Should we just say `first PR review`. I'm thinking that even a small review prevent the use of force-push, no ?

Reviewer 4: ➕, first review should prevent usage of force push

Author: I stuck to what was discussed but I agree that first review should prevent usage of force push.

> #### Before the first in-depth PR review

Reviewer 4: Should we explicitly recommend to mark a PR as draft before it's ready for a review?

Reviewer 3: I don't think so. In real life you simply don't open the PR until you're ready to receive reviews.

Only long running PR, where you want early feedback, get into this situation.

Reviewer 5: I don't think there's disagreement here actually :)

I think we can add a recommendation that PRs should only be opened once they're reviewable, that _draft_ PRs should be used to share code before it's ready for actual reviews.

Before the first in-depth PR review, meaningful commits are best: logically-encapsulated commits help the reviews go quicker and make the job for the reviewer easier. Conflicts with `main` can be resolved with a `git rebase origin/main` and a force push if it makes the review easier.

#### After the first in-depth review

After the first in-depth review, to make follow-up reviews easier:

> - Avoid force pushes (in particular rewriting the history that was already reviewed, which makes follow up reviews painful)

Reviewer 3: I think we should explain why and how it's linked to GH. People familiar with Git but not GH might find it weird otherwise.

WDYT:
```suggestion
- Avoid force pushes. In particular rewriting the history that was already reviewed, which makes follow up reviews painful as GitHub loose track of each comments.
```

Reviewer 6: Small edit: "loses" instead of "loose"

> - Resolve merge conflicts with `main` using `git merge origin/main` (AFAIK Github handles that well in the history and the diff)

Reviewer 4: I don't think we need the "AFAIK" comment here

```suggestion
- Resolve merge conflicts with `main` using `git merge origin/main`
```

#### How to merge to `main`

Once reviews are complete, the merge to `main` should be done with either:
- the squash-merge option, to keep the history of `main` clean (even though some context/details are lost in the squash). The commit message for this squash should always be edited to concisely describe the commit without extraneous “address review comments” text. 

> - the “rebase-merge” option, after manually rewriting the PR’s commit history and force-pushing to the branch. 

Reviewer: We should note a bit more explicitly that a rebase-merge is only appropriate with a clean history in the branch

complete no-op for the test suite, please add **[skip ci]** in the commit message
body to skip the build and give that slot to someone else who does need it.

### Pull request workflow

The goals ordered by priority are:
- Make PR reviews (both initial and follow-up reviews) easy for reviewers using
 GitHub
- On the `main` branch, have a meaningful commit history that allows
 understanding (even years later) what each commit does, and why. 

You must open the PR when the code is reviewable or you must set the PR as
 draft if you want to share code before it's ready for actual reviews.

#### Before the first PR review

Before the first PR review, meaningful commits are best: logically-encapsulated
 commits help the reviews go quicker and make the job for the reviewer easier.
 Conflicts with `main` can be resolved with a `git rebase origin/main` and a

>  force push if it makes the review easier.

Reviewer 5: ```suggestion
 force push if it makes future review(s) easier.
```

#### After the first review

After the first review, to make follow-up reviews easier:

> - Avoid force pushes. In particular rewriting the history that was already
>  reviewed, which makes follow up reviews painful as GitHub loses track of each
>  comments.

Reviewer 5: ```suggestion
- Avoid force pushes: rewriting the history that was already
 reviewed makes follow-up reviews painful as GitHub loses track of each
 comment. Instead, address reviews with additional commits on the PR branch.
```
